### ISSUE TITLE
Facing 422 issue while making POST and DELETE calls.

### ISSUE THREAD

**Issue by VenkatRamReddyK** [stoplightio/prism#1033]
**Describe the bug**
We use Prism Mock Server for development and mock server for running API contracts tests for our React application. With the recent update 3.3.0 we figured out some of the tests were failing.
As a practice, we removed node modules, did npm cache clean --force and restarted prism and ran tests.

**To Reproduce**

Here is the yaml code for simulating 422 issue

**YAML code PRISM Uses** 

```
   /api/users: 
      post: 
        summary: "Create a user"
        operationId: "dal.users.create_and_get"
        requestBody: 
          description: "Parameters to define a new user."
          content: 
            application/json: 
              schema: 
                x-body-name: "user"
                $ref: "#/components/schemas/UserCreateEntity"
        responses: 
          200: 
            description: "User created successfully"
            content: 
              application/json: 
                schema: 
                  $ref: "#/components/schemas/UserCreateResponseEntity"
          409: 
            description: "Duplicate user name"
      get: 
        summary: "List all users for a tenant"
        operationId: "dal.users.get_all"
        responses: 
          200: 
            description: "Get a list of all Users "
            content: 
              application/json: 
                schema: 
                  $ref: "#/components/schemas/UserCollectionEntity"

    /api/users/{user_id}: 
      delete: 
        summary: "Delete a user"
        operationId: "dal.users.delete"
        parameters: 
          - 
            name: "user_id"
            in: "path"
            description: "Id of User"
            required: true
            schema: 
              $ref: "#/components/schemas/UserId"
        responses: 
          204: 
            description: "User deleted successfully"
```

**Redux actions consuming prism response** 

```
it('should dispatch addUserAction action to the store', (done) => {

    store = mockStore({});

    let params = {
      name: 'test'
    };

    store.dispatch(userActions.addUserAction(params))
      .then((orchestrationResponse) => {
        // The below snippet tests the POST `/api/users` call against the API contract
        const addUseraction = store.getActions()[0];
        const toggleUserRegAction = store.getActions()[1];

        expect(addUseraction.type).toEqual('ADD_USER');

        expect(addUseraction).toHaveProperty('payload');
        expect(addUseraction.payload).toHaveProperty('id');
        expect(addUseraction.payload).toHaveProperty('name');
        expect(addUseraction.payload).toHaveProperty('appliance_version');

        done();
      });

  });

  it('should dispatch removeUserAction action to the store', (done) => {

    store = mockStore({});

    //Loading all the users
    store.dispatch(userActions.fetchUsersAction())
      .then(() => {
        const users = store.getActions()[0].payload;
        expect(store.getActions()[0].type).toEqual('LOAD_USERS');

        if (users.length > 0) {
          store.dispatch(userActions.removeUserAction(users[0]))
            .then(() => {
        // The below snippet tests the DELETE `/api/users` call against the API contract
              expect(store.getActions()[1].type).toEqual('REMOVE_USER');
              done();
            });
        }
        else {
          done();
        }

      });

  });
```

**These are the errors we are experiencing when adding and deletion User action**

```
  ● ● test users async actions › should dispatch removeUserAction action to the store

    Request failed with status code 422

      1109 |  */
      1110 | var createError = function createError(message, config, code, request, response) {
    > 1111 |   var error = new Error(message);
           |               ^
      1112 |   return enhanceError(error, config, code, request, response);
      1113 | };
      1114 | 

      at createError (../shared-components/dist/bundle.js:1111:15)
      at settle (../shared-components/dist/bundle.js:1128:12)
      at XMLHttpRequest.handleLoad (../shared-components/dist/bundle.js:1415:7)
      at XMLHttpRequest.<anonymous> (../../node_modules/jsdom/lib/jsdom/living/helpers/create-event-accessor.js:33:32)
      at invokeEventListeners (../../node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:193:27)
      at XMLHttpRequestEventTargetImpl._dispatch (../../node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:119:9)
      at XMLHttpRequestEventTargetImpl.dispatchEvent (../../node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:82:17)
      at XMLHttpRequest.dispatchEvent (../../node_modules/jsdom/lib/jsdom/living/generated/EventTarget.js:157:21)
      at readyStateChange (../../node_modules/jsdom/lib/jsdom/living/xmlhttprequest.js:811:9)
      at Request.<anonymous> (../../node_modules/jsdom/lib/jsdom/living/xmlhttprequest.js:939:7)
      at IncomingMessage.<anonymous> (../../node_modules/request/request.js:1076:12)

  ● test users async actions › should dispatch addUserAction action to the store

    Request failed with status code 422

      1109 |  */
      1110 | var createError = function createError(message, config, code, request, response) {
    > 1111 |   var error = new Error(message);
           |               ^
      1112 |   return enhanceError(error, config, code, request, response);
      1113 | };
      1114 | 

      at createError (../shared-components/dist/bundle.js:1111:15)
      at settle (../shared-components/dist/bundle.js:1128:12)
      at XMLHttpRequest.handleLoad (../shared-components/dist/bundle.js:1415:7)
      at XMLHttpRequest.<anonymous> (../../node_modules/jsdom/lib/jsdom/living/helpers/create-event-accessor.js:33:32)
      at invokeEventListeners (../../node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:193:27)
      at XMLHttpRequestEventTargetImpl._dispatch (../../node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:119:9)
      at XMLHttpRequestEventTargetImpl.dispatchEvent (../../node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:82:17)
      at XMLHttpRequest.dispatchEvent (../../node_modules/jsdom/lib/jsdom/living/generated/EventTarget.js:157:21)
      at readyStateChange (../../node_modules/jsdom/lib/jsdom/living/xmlhttprequest.js:811:9)
      at Request.<anonymous> (../../node_modules/jsdom/lib/jsdom/living/xmlhttprequest.js:939:7)
      at IncomingMessage.<anonymous> (../../node_modules/request/request.js:1076:12)

```

**Environment**

Library version: @stoplight/prism-cli": "3.3.0",
OS: MacOS Mojave
Browser: Chrome 78
Additional context

**Note**: I tried to maintain 1 endpoint with  just POST and I can simulate the error as well. So this issue is a new one unlike #981 . get users seems to be working good.

**Comment by XVincentX** [comment]
Hey @VenkatRamReddyK 

Thanks for the message!

the 422 error code is returned by Prism because of a validation error of the input — which you didn't show in this issue. To fully debut this, I would need:

1. The complete OAS document (the one you provided has references and I can't tell what's in there)
2. The sample input you're sending to the server

**Comment by VenkatRamReddyK** [comment]
@XVincentX, 

I see the root cause of this issue is **prism@3.3.0 is not responding with dynamic data and returning data from example even after using --dynamic flag**.
I am running prism mock server using `prism mock masterSpec.inc.yaml --dynamic`

The following are my findings along with the changes to make the tests work. 
And as per your ask, I added working yaml file which has **All Tests Pass** after my changes.

**Finding**

For testing this I used postman for DELETE, POST, GET calls for Users and UserWorkspaces just to ensure if both the versions are responding with correct data. Followed by updating the failed test-cases accordingly.

**prism@3.3.0 is not responding with dynamic data**
"@stoplight/prism-cli": "3.2.9" version sends dynamic data while 3.3.0 doesn't with the following way. `prism mock masterSpec.inc.yaml --dynamic`
For 3.3.0, if data is not dynamic, prism responds with the example data. and if the example data doesn't have any required fields it throws the error with 422. i.e. **Request failed with status code 422**. As a fix, I tried to add all the required fields adhere to constraints which threw error in the logs when I ran prism.  This fixed 1 issue for addUserAction. 
For removeUserAction, we are doing a getALLUsers and passing the first user response as the request params, which has id. Ideally it should work. But the response from getALLUsers is from example which is not as per the norms of UUID. So, when I was playing with postman, I realized if we rend a valid UUID  as params removeUserAction should work. Hence, fixed it.  
The following is the change I did.

```
it('should dispatch removeUserAction action to the store', (done) => {

    store = mockStore({});
    let userParams = {
      id: 'd7d155ec-dee9-d771-e9fd-f84d62e6b59a'
    };
    let UUID="^([0-9a-fA-F]{8})-(([0-9a-fA-F]{4}\-){3})([0-9a-fA-F]{12})$";
    expect(userParams.id.match(UUID)).not.toBe(null);
    store.dispatch(userActions.removeUserAction(userParams))
    .then(() => {
      expect(store.getActions()[0].type).toEqual('REMOVE_USER');
      done();
    });

  });
```

1. The Request and response to the api endpoint. (Request we are sending to api endpoint is in the request params)

a) **Request URL for POST Users**:/api/users?name=test&location[name]=US


{
    "id": "d7d155ec-dee9-d771-e9fd-f84d62e6b59a",
    "name": "test_user",
    "state": "unregistered",
    "created": "2020-03-12T06:51:50Z",
    "updated": "2020-03-12T06:51:50Z",
    "type": null,
    "model": null,
    "serial_number": null,
    "device_version": null,
    "sdp_mode": "user",
    "notification_channel_status": "offline",
    "workspaces": {
        "type": "vsphere",
        "mode": "manual"
    }
}

b) **Request URL for GET Users**:  /api/users

**Response:**

```
[
    {
        "id": "e274bf3ebe3841a88ade1630515624c6",
        "name": "string",
        "state": "unregistered",
        "created": "2020-03-12T15:59:28Z",
        "updated": "2020-03-12T15:59:28Z",
        "type": null,
        "model": null,
        "serial_number": null,
        "device_version": null,
        "sdp_mode": "pzt-user",
        "location": {
            "name": null
        },
        "notification_channel_status": "online",
        "workspaces": {
            "type": "ios",
            "mode": "auto",
            "state": "waiting-to-create"
        },
        "external_ip": null,
        "external_fqdn": "string",
        "public_ip": "string"
    }
]
```

c) **DELETE Request URL for Users**:  /api/users/d7d155ec-dee9-d771-e9fd-f84d62e6b59a

**Response**

```
no response as it is a 204
```

d) **POST Request Users for workspaces**: /api/users/85d9463a-eca8-82fa-712b-019302bf597f/workspaces

**Response**

```
{
    "device_id": "e274bf3ebe3841a88ade1630515624c6",
    "device_config": {
        "admin_username": "string",
        "internal_fqdn": "string",
        "external_fqdn": "string",
        "management_fqdn": "string",
        "company_name": "string",
        "primary_dns": "string",
        "secondary_dns": "string",
        "private_domain_name": "string",
        "public_domain_name": "string",
        "management_domain_name": "string",
        "internal_ip_address": "string",
        "internal_subnet": "string",
        "internal_user": "string",
        "external_ip_address": "string",
        "external_subnet": "string",
        "external_user": "string",
        "management_ip_address": "string",
        "public_ip_address": "string",
        "management_subnet": "string",
        "management_user": "string",
        "dns_search_domain": "string"
    }
}
```

2. As per your ask below is the master yaml spec which is a combination of 2 yaml files. 
 **Source: openapi/masterSpec.yaml**

```
openapi: 3.0.0
info:
  title: User Workspace Config API
  version: '0.1'
components:
  schemas:
    UserId:
      type: string
      format: uuid
      description: Unique device identifier hex uuid string
      example: e274bf3ebe3841a88ade1630515624c6
    UserCreateEntity:
      type: object
      required:
        - name
        - workspaces
      properties:
        name:
          type: string
          description: user name
        workspaces:
          description: Workspace attributes. Only `type` is required for user creation.
          $ref: '#/components/schemas/WorkspacedDeviceEntity'
        location:
          $ref: '#/components/schemas/UserLocationEntity'
    UserCreateResponseEntity:
      $ref: '#/components/schemas/UserEntity'
    UserEntity:
      type: object
      required:
        - id
        - name
        - state
        - created
        - updated
        - type
        - model
        - serial_number
        - device_version
        - notification_channel_status
      properties:
        id:
          $ref: '#/components/schemas/UserId'
        name:
          type: string
          description: Device name
        state:
          type: string
          description: Device state
          enum:
            - unregistered
            - registered
            - deleted
        created:
          type: string
          format: date-time
          description: UTC RFC 3339 timestamp of when device is created
        updated:
          type: string
          format: date-time
          description: UTC RFC 3339 timestamp of when device is updated
        type:
          type: string
          description: Device type
          nullable: true
        model:
          type: string
          description: Device model
          nullable: true
        serial_number:
          type: string
          description: Device serial number
          nullable: true
        device_version:
          type: string
          description: Device version
          nullable: true
        sdp_mode:
          type: string
          description: Device SDP mode
          enum:
            - pzt-user
          nullable: true
        location:
          $ref: '#/components/schemas/UserLocationEntity'
        notification_channel_status:
          type: string
          description: Status of the notification channel
          enum:
            - online
            - offline
          nullable: true
        workspaces:
          $ref: '#/components/schemas/WorkspacedDeviceEntity'
        external_ip:
          type: string
          nullable: true
          description: External IP address of the user
        external_fqdn:
          type: string
          description: External FQDN address of the user
        public_ip:
          type: string
          description: Public IP address of the user
    WorkspacedDeviceEntity:
      type: object
      required:
        - type
      properties:
        type:
          type: string
          description: >-
            workspaces type of the device. If type is not presented, the user is
            not workspaced.
          enum:
            - ios
            - android
            - windows
        mode:
          type: string
          description: workspaces mode of the device
          enum:
            - auto
            - manual
          nullable: true
        state:
          type: string
          description: workspaces state of the device
          enum:
            - waiting-to-create
            - waiting-to-update
            - waiting-to-start
            - waiting-to-stop
            - waiting-to-destroy
            - creating
            - updating
            - paused
            - starting
            - started
            - stopping
            - destroying
            - destroyed
            - failed-to-create
            - failed-to-destroy
            - waiting-on-certificate
            - configuring
            - configured
          nullable: true
    UserCollectionEntity:
      type: array
      items:
        $ref: '#/components/schemas/UserEntity'
    UserLocationEntity:
      type: object
      properties:
        name:
          type: string
          description: Custom name of the location
          nullable: true
    UserLocationsEntity:
      type: object
      properties:
        names:
          type: array
          description: Existing custom location names
          items:
            type: string
    WorkspaceDeviceConfigBaseEntity:
      type: object
      properties:
        admin_username:
          type: string
          description: Admin username of the user
        internal_fqdn:
          type: string
          description: Internal FQDN of the user
        external_fqdn:
          type: string
          description: External FQDN of the user
        management_fqdn:
          type: string
          description: Management FQDN of the user
        company_name:
          type: string
          description: Company name of the user
        primary_dns:
          type: string
          description: Primary DNS of the user
        secondary_dns:
          type: string
          description: Secondary DNS of the user
        private_domain_name:
          type: string
          description: Private Domain name of the user
        public_domain_name:
          type: string
          description: Public domain name of the user
        management_domain_name:
          type: string
          description: Management domain name of the user
        internal_ip_address:
          type: string
          description: Internal IP address of the user
        internal_subnet:
          type: string
          description: Internal subnet of the user
        internal_user:
          type: string
          description: Internal user of the user
        external_ip_address:
          type: string
          description: External IP address of the user
        external_subnet:
          type: string
          description: External subnet of the user
        external_user:
          type: string
          description: External user of the user
        management_ip_address:
          type: string
          description: Management IP address of the user
        public_ip_address:
          type: string
          description: The IP address for which a DNS record will be created
        management_subnet:
          type: string
          description: Mangement subnet of the user
        management_user:
          type: string
          description: Mangement user of the user
        dns_search_domain:
          type: string
          description: User's DNS search domain.
    WorkspaceDeviceConfigCreateEntity:
      allOf:
        - $ref: '#/components/schemas/WorkspaceDeviceConfigBaseEntity'
        - type: object
          properties:
            admin_password:
              type: string
              description: Admin password of the user
    WorkspaceDeviceConfigEntity:
      $ref: '#/components/schemas/WorkspaceDeviceConfigBaseEntity'
    WorkspaceConfigCreateEntity:
      type: object
      properties:
        device_config:
          $ref: '#/components/schemas/WorkspaceDeviceConfigCreateEntity'
    WorkspaceConfigEntity:
      type: object
      properties:
        device_id:
          $ref: '#/components/schemas/UserId'
        device_config:
          $ref: '#/components/schemas/WorkspaceDeviceConfigEntity'
    WorkspaceConfigResponseEntity:
      $ref: '#/components/schemas/WorkspaceConfigEntity'
paths:
  /api/users:
    post:
      summary: Create a user
      operationId: dal.users.create_and_get
      requestBody:
        description: Parameters to define a new user.
        content:
          application/json:
            schema:
              x-body-name: user
              $ref: '#/components/schemas/UserCreateEntity'
            example:
              name: test_user
              workspaces_type: vsphere
      responses:
        '200':
          description: User created successfully
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/UserCreateResponseEntity'
              example:
                id: d7d155ec-dee9-d771-e9fd-f84d62e6b59a
                name: test_user
                state: unregistered
                created: '2020-03-12T06:51:50Z'
                updated: '2020-03-12T06:51:50Z'
                type: null
                model: null
                serial_number: null
                device_version: null
                sdp_mode: user
                notification_channel_status: offline
                workspaces:
                  type: vsphere
                  mode: manual
        '409':
          description: Duplicate user name
    get:
      summary: List all users.for a tenant
      operationId: dal.users.get_all
      responses:
        '200':
          description: Get a list of all Users for the tenant.
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/UserCollectionEntity'
  '/api/users/{user_id}':
    get:
      summary: Get details of a user by id
      operationId: dal.users.get
      parameters:
        - name: user_id
          in: path
          description: Id of User
          required: true
          schema:
            $ref: '#/components/schemas/UserId'
      responses:
        '200':
          description: User retrieved successfully
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/UserEntity'
        '404':
          description: User not found
    delete:
      summary: Delete a user
      operationId: dal.users.delete
      parameters:
        - name: user_id
          in: path
          description: Id of User
          required: true
          schema:
            $ref: '#/components/schemas/UserId'
      responses:
        '204':
          description: User deleted successfully
  /api/users/locations:
    get:
      summary: >-
        Get location info of the users. It returns a list of existing custom
        name.
      operationId: dal.users.get_locations
      responses:
        '200':
          description: Retrieve location info successfully
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/UserLocationsEntity'
  '/api/users/{user_id}/workspaces':
    post:
      summary: Create user's workspaces config
      operationId: dal.workspaces_config.create_and_get
      parameters:
        - name: user_id
          in: path
          description: Id of User
          required: true
          schema:
            $ref: '#/components/schemas/UserId'
      requestBody:
        description: Parameters to define user's workspaces config.
        content:
          application/json:
            schema:
              x-body-name: workspaces_config
              $ref: '#/components/schemas/WorkspaceConfigCreateEntity'
      responses:
        '200':
          description: User workspaces config created successfully
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/WorkspaceConfigResponseEntity'
        '404':
          description: Device ID is invalid.
```

**Comment by pytlesk4** [comment]
@VenkatRamReddyK can you try using the [Prefer Header](https://github.com/stoplightio/prism/blob/master/docs/guides/01-mocking.md#static-or-dynamic-generation) for dynamic mocking? 

While we figure out what to do about the dynamic mocking flag. I feel like this is a bug and we should just automatically add the Prefer header when a request comes into prism like so: `Prefer: dyanmic=true`.

**Comment by VenkatRamReddyK** [comment]
`curl http://127.0.0.1:4010/api/users -H "Prefer: dynamic=true"` works for us. 
Since, 3.2.9 works for us. We will wait for your fix before upgrading to 3.3.0.

**Comment by XVincentX** [comment]
Most likely I put a bug where the un-presence of the header is taken as false and not properly overridden. We'll fix this soon and release 3.3.1
